c_src/sqlite3_nif.c

```diff
@@ -1065,6 +1068,69 @@ exqlite_set_update_hook(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
     return make_atom(env, "ok");
 }
 
+//
+// Log Notifications
+//
+
+void
+log_callback(void* arg, int iErrCode, const char* zMsg)
+{
+    if (log_hook_pid == NULL) {
+        return;
+    }
+
+    ErlNifEnv* msg_env = enif_alloc_env();
+    ERL_NIF_TERM error = make_binary(msg_env, zMsg, strlen(zMsg));
+    ERL_NIF_TERM msg   = enif_make_tuple3(msg_env, make_atom(msg_env, "log"), enif_make_int(msg_env, iErrCode), error);
+
+    if (!enif_send(NULL, log_hook_pid, msg_env, msg)) {
+        enif_mutex_lock(log_hook_mutex);
+        sqlite3_config(SQLITE_CONFIG_LOG, NULL, NULL);
+        enif_free(log_hook_pid);
+        log_hook_pid = NULL;
+        enif_mutex_unlock(log_hook_mutex);
+    }
+
+    enif_free_env(msg_env);
+}
+
+static ERL_NIF_TERM
+exqlite_set_log_hook(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+{
+    assert(env);
+
+    if (argc != 1) {
+        return enif_make_badarg(env);
+    }
+
+    ErlNifPid* pid = (ErlNifPid*)enif_alloc(sizeof(ErlNifPid));
+    if (!enif_get_local_pid(env, argv[0], pid)) {
+        enif_free(pid);
+        return make_error_tuple(env, "invalid_pid");
+    }
+
+    if (log_hook_mutex == NULL) {
+        log_hook_mutex = enif_mutex_create("exqlite:log_hook");
+
+        if (log_hook_mutex == NULL) {
+            return make_error_tuple(env, "failed_to_create_mutex");
+        }
+    }
```

Comment:
I believe this still has a chance of a race condition. This should be chucked into the `on_load` function and then the teardown of the mutex thrown into `on_unload`.

Reply:
Do you mean multiple threads calling `enif_mutex_create`? It's marked as thread safe in the docs.

Reply:
Outside of this, I am curious what the behavior is when two threads call `enif_mutex_create` with the same parameter. Obviously the one that called it first will win, but what happens to the loser?

Reply:
Ah, it probably should've been

```c
ErlNifMutex* mutex = enif_mutex_create("exqlite:log_hook");

if (mutex) {
  log_hook_mutes = mutex;
}  else {
  return make_error_tuple(env, "failed_to_create_mutex");
}
```

Reply:
Moved mutex to on_load in https://github.com/elixir-sqlite/exqlite/pull/266/commits/dc138b2744f8ad9fbbdf488ba3a5b7f39338926c